validation/tests/test_scenario_generator.py

```diff
@@ -23,101 +25,382 @@ def _get_dataset():
     latest_df = pd.read_csv(DATA_FILE,
                             parse_dates=['Date'],
                             encoding="ISO-8859-1",
+                            dtype={"RegionName": str,
+                                   "RegionCode": str},
                             error_bad_lines=False)
     latest_df["RegionName"] = latest_df["RegionName"].fillna("")
+    # Fill any missing NPIs by assuming they are the same as previous day, or 0 if none is available
+    latest_df.update(latest_df.groupby(['CountryName', 'RegionName'])[NPI_COLUMNS].ffill().fillna(0))
     return latest_df
 
 
 class TestScenarioGenerator(unittest.TestCase):
+    """
+    Tests generating different NPI scenarios.
 
-    def test_generate_scenario_historical(self):
-        latest_df = _get_dataset()
-        start_date_str = "2020-08-01"
-        end_date_str = "2020-08-4"
+    Definitions:
+    I = inception date = 2020-01-01 (earliest available data)
+    LK = last known date (for each country) in the latest available data
+    S = start date of the scenario
+    E = end date of the scenario
+
+    Time wise, the following kind of scenarios can be applied:
+    1. Counterfactuals: I____S_____E____LK where E can equal LK
+    2. Future:          I____S_____LK___E where S can equal LK
+    3. Mind the gap:    I____LK    S____E
+
+    For each case, we check each type of scenario: freeze, MIN, MAX, custom
+
+    Scenarios can be applied to: 1 country, several countries, all countries
```

Comment:
These are the cases we have to handle, I think. The tests below are not completely rational, but I think they do the job.